File: oauth2-server/src/main/java/com/clouway/oauth2/token/JjwtIdTokenFactory.java
package com.clouway.oauth2.token;

import com.clouway.oauth2.DateTime;
import com.clouway.oauth2.Identity;
import com.clouway.oauth2.KeyStore;
import com.clouway.oauth2.client.IdentityKeyPair;
import com.google.common.base.Optional;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;

import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

/**
 * @author Vasil Mitov <vasil.mitov@clouway.com>
 */
public class JjwtIdTokenFactory implements IdTokenFactory {
  private final KeyStore keyStore;

  public JjwtIdTokenFactory(KeyStore keyStore) {
    this.keyStore = keyStore;
  }

  @Override
  public Optional<String> create(String host, String clientId, Identity identity, Long ttl, DateTime instant) {
    List<IdentityKeyPair> keys = keyStore.getKeys();

    if (keys == null || keys.isEmpty()) {
      return Optional.absent();
    }

    IdentityKeyPair signingKey = randomKey(keys);

    Map<String, Object> claims = new LinkedHashMap<>();
    claims.put("iss", host);
    claims.put("aud", clientId);
    claims.put("sub", identity.id());
    claims.put("name", identity.name());
    claims.put("email", identity.email());
    claims.put("given_name", identity.givenName());
    claims.put("family_name", identity.familyName());
    claims.putAll(identity.claims());

    return Optional.of(Jwts.builder()
            .setHeaderParam("cid", signingKey.keyId)//CertificateId - the ID of the certificate that the token was signed with.
            .setClaims(claims)
            .setIssuedAt(new Date(instant.timestamp()))
            .setExpiration(new Date(instant.timestamp() + ttl))
            .signWith(SignatureAlgorithm.RS256, signingKey.privateKey)
            .compact());
  }

  private IdentityKeyPair randomKey(List<IdentityKeyPair> keys) {
    Random random = new Random();
    return keys.get(random.nextInt(keys.size()));
  }

}

File: oauth2-server/src/main/java/com/clouway/oauth2/token/JjwtIdTokenFactory.kt
package com.clouway.oauth2.token

import com.clouway.oauth2.common.DateTime
import com.clouway.oauth2.keystore.IdentityKeyPair
import com.clouway.oauth2.keystore.KeyStore
import com.clouway.oauth2.util.Hashes
import io.jsonwebtoken.Jwts
import java.util.Date
import java.util.Random

/**
 * @author Vasil Mitov <vasil.mitov></vasil.mitov>@clouway.com>
 */
class JjwtIdTokenFactory(
    private val keyStore: KeyStore,
) : IdTokenFactory {
    override fun newBuilder(): IdTokenBuilder = DefaultIdTokenBuilder(keyStore)
}

private class DefaultIdTokenBuilder(
    private val keyStore: KeyStore,
) : IdTokenBuilder {
    private val claims: MutableMap<String, Any> = LinkedHashMap()
    private var issuedAt: DateTime? = null
    private var ttl: Long? = null

    override fun issuer(issuer: String): IdTokenBuilder {
        claims["iss"] = issuer
        return this
    }

    override fun audience(audience: String): IdTokenBuilder {
        claims["aud"] = audience
        return this
    }

    override fun issuedAt(instant: DateTime): IdTokenBuilder {
        this.issuedAt = instant
        return this
    }

    override fun ttl(ttlSeconds: Long): IdTokenBuilder {
        this.ttl = ttlSeconds
        return this
    }

    override fun subjectUser(identity: Identity): IdTokenBuilder {
        claims["sub"] = identity.id
        claims["subject_type"] = "user"
        claims["name"] = identity.name
        claims["email"] = identity.email
        claims["given_name"] = identity.givenName
        claims["family_name"] = identity.familyName
        claims.putAll(identity.claims)
        return this
    }

    override fun subjectServiceAccount(serviceAccount: ServiceAccount): IdTokenBuilder {
        claims["sub"] = serviceAccount.clientId
        claims["subject_type"] = "service_account"
        claims["name"] = serviceAccount.name
        claims["email"] = serviceAccount.clientEmail
        claims["given_name"] = ""
        claims["family_name"] = ""
        claims.putAll(serviceAccount.claims)
        return this
    }

    override fun withAccessToken(accessToken: String): IdTokenBuilder {
        claims["at_hash"] = Hashes.atHash(accessToken)
        return this
    }

    override fun claim(
        name: String,
        value: Any,
    ): IdTokenBuilder {
        claims[name] = value
        return this
    }

    override fun build(): String {
        val keys = keyStore.keys ?: throw IllegalStateException("No signing keys are configured")
        if (keys.isEmpty()) throw IllegalStateException("No signing keys are configured")

        val iat = issuedAt ?: throw IllegalStateException("issuedAt must be specified")
        val ttlSeconds = ttl ?: throw IllegalStateException("ttl must be specified")

        if (!claims.containsKey("sub")) {
            throw IllegalStateException("subject must be specified")
        }

        val signingKey = selectRandomKey(keys)
        val jwt =
            Jwts
                .builder()
                .header()
                .keyId(signingKey.keyId)
                .and()
                .claims(claims)
                .issuedAt(Date(iat.timestamp()))
                .expiration(Date(iat.timestamp() + ttlSeconds))
                .signWith(signingKey.privateKey, Jwts.SIG.RS256)
                .compact()
        return jwt
    }
}

private fun selectRandomKey(keys: List<IdentityKeyPair>): IdentityKeyPair {
    val random = Random()
    return keys[random.nextInt(keys.size)]
}
